execution_engine/src/storage/trie_store/lmdb.rs

```diff
@@ -161,3 +172,150 @@ impl<K, V> Store<Digest, Trie<K, V>> for LmdbTrieStore {
 }
 
 impl<K, V> TrieStore<K, V> for LmdbTrieStore {}
+
+pub(crate) type Cache = Arc<Mutex<HashMap<Digest, (bool, Trie<Key, StoredValue>)>>>;
+/// In-memory cached trie store, backed by rocksdb.
+#[derive(Clone)]
+pub struct ScratchCache {
+    pub(crate) cache: Cache,
+    pub(crate) store: Arc<LmdbTrieStore>,
+    pub(crate) env: Arc<LmdbEnvironment>,
+}
+
+/// Cached version of the trie store.
+#[derive(Clone)]
+pub struct ScratchTrieStore {
+    pub(crate) inner: ScratchCache,
+}
+
+impl ScratchTrieStore {
+    /// Creates a new ScratchTrieStore.
+    pub fn new(store: Arc<LmdbTrieStore>, env: Arc<LmdbEnvironment>) -> Self {
+        Self {
+            inner: ScratchCache {
+                store,
+                env,
+                cache: Default::default(),
+            },
+        }
+    }
+
+    /// Writes all dirty (modified) cached tries to underlying db.
+    pub fn write_all_tries_to_db(self, new_state_root: Digest) -> Result<(), error::Error> {
+        let env = self.inner.env;
+        let store = self.inner.store;
+        let cache = &mut *self.inner.cache.lock().map_err(|_| error::Error::Poison)?;
+
+        let mut missing_trie_keys = vec![new_state_root];
+        let mut validated_tries = HashMap::new();
+
+        let mut txn = env.create_read_write_txn()?;
+
+        while let Some(next_trie_key) = missing_trie_keys.pop() {
+            if cache.is_empty() {
+                return Err(error::Error::CommitError(
+                    CommitError::TrieNotFoundDuringCacheValidate(next_trie_key),
+                ));
+            }
+            match cache.remove(&next_trie_key) {
+                Some((false, _)) => continue,
+                None => {
+                    txn.read(store.get_db(), next_trie_key.as_ref())?.ok_or(
+                        error::Error::CommitError(CommitError::TrieNotFoundDuringCacheValidate(
+                            next_trie_key,
+                        )),
+                    )?;
```

Comment:
Not sure if this is worthwhile?  Maybe we should just skip the check (since it would only happen due to programmer error or DB corruption I guess?) as the missing trie might not be a fatal problem, but would result in a fatal error here.

Reply:
Yeah it doesn't look like this check is necessary I agree

Reply:
resolved in [4b78471](https://github.com/casper-network/casper-node/pull/2877/commits/4b78471c0fbac1bcaf8999792dcda1d0ded85d70)